File: AssetManager/Helpers/OtherFunctions.vb
Imports System.IO
Imports MyDialogLib

Module OtherFunctions
    Public stpw As New Stopwatch
    Private intTimerHits As Integer = 0

    Public Sub StartTimer()
        stpw.Stop()
        stpw.Reset()
        stpw.Start()
    End Sub

    Public Function StopTimer() As String
        stpw.Stop()
        intTimerHits += 1
        Dim Results As String = intTimerHits & "  Stopwatch: MS:" & stpw.ElapsedMilliseconds & " Ticks: " & stpw.ElapsedTicks
        Debug.Print(Results)
        Return Results
    End Function

    Public Function ElapTime() As String
        Dim results As String = intTimerHits & "  Elapsed: MS:" & stpw.ElapsedMilliseconds & " Ticks: " & stpw.ElapsedTicks
        Debug.Print(results)
        Return results
    End Function

    Public Sub EndProgram()
        GlobalSwitches.ProgramEnding = True
        Logger("Ending Program...")
        PurgeTempDir()
        Application.Exit()
    End Sub

    Public Sub PurgeTempDir()
        Try
            Directory.Delete(Paths.DownloadPath, True)
        Catch
        End Try
    End Sub

    Public Sub AdjustComboBoxWidth(ByVal sender As Object, ByVal e As EventArgs)
        Dim senderComboBox = DirectCast(sender, ComboBox)
        Dim width As Integer = senderComboBox.DropDownWidth
        Dim g As Graphics = senderComboBox.CreateGraphics()
        Dim font As Font = senderComboBox.Font
        Dim vertScrollBarWidth As Integer = If((senderComboBox.Items.Count > senderComboBox.MaxDropDownItems), SystemInformation.VerticalScrollBarWidth, 0)
        Dim newWidth As Integer
        For Each s As String In DirectCast(sender, ComboBox).Items
            newWidth = CInt(g.MeasureString(s, font).Width) + vertScrollBarWidth
            If width < newWidth Then
                width = newWidth
            End If
        Next
        senderComboBox.DropDownWidth = width
    End Sub

    Public Function NotePreview(Note As String, Optional CharLimit As Integer = 50) As String
        If Note <> "" Then
            Return Strings.Left(Note, CharLimit) & IIf(Len(Note) > CharLimit, "...", "").ToString
        Else
            Return ""
        End If
    End Function

    Public Function Message(ByVal Prompt As String, Optional ByVal Buttons As Integer = vbOKOnly + vbInformation, Optional ByVal Title As String = Nothing, Optional ByVal ParentFrm As Form = Nothing) As MsgBoxResult
        SetWaitCursor(False, ParentFrm)
        Dim NewMessage As New AdvancedDialog(ParentFrm)
        Return NewMessage.DialogMessage(Prompt, Buttons, Title, ParentFrm)
    End Function

    Public Function OKToEnd() As Boolean
        If GlobalSwitches.BuildingCache Then
            Message("Still building DB Cache. Please wait and try again.", vbOKOnly + vbInformation, "Critical Function Running")
            Return False
        End If
        If GKUpdaterForm.Visible AndAlso Not GKUpdaterForm.OkToClose() Then Return False
        Return True
    End Function

    Public Sub SetWaitCursor(Waiting As Boolean, Optional parentForm As Form = Nothing)
        If parentForm Is Nothing Then
            Application.UseWaitCursor = Waiting
            Application.DoEvents()
        Else
            If Waiting Then
                parentForm.Cursor = Cursors.AppStarting
            Else
                parentForm.Cursor = Cursors.Default
            End If
            parentForm.Update()
        End If

    End Sub

    Public Function RTFToPlainText(rtfText As String) As String
        Try
            If rtfText.StartsWith("{\rtf") Then
                Using rtBox As New RichTextBox
                    rtBox.Rtf = rtfText
                    Return rtBox.Text
                End Using
            Else
                Return rtfText
            End If
        Catch ex As ArgumentException
            'If we get an argument error, that means the text is not RTF so we return the plain text.
            Return rtfText
        End Try
    End Function

    Public Sub SetRichTextBox(ByRef richTextBox As RichTextBox, text As String)
        If text.StartsWith("{\rtf") Then
            richTextBox.Rtf = text
        Else
            richTextBox.Text = text
        End If
    End Sub

End Module

File: AssetManager/Helpers/OtherFunctions.cs
using System;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.IO;
using MyDialogLib;
namespace AssetManager
{

    static class OtherFunctions
    {
        public static Stopwatch stpw = new Stopwatch();

        private static int intTimerHits = 0;
        public static void StartTimer()
        {
            stpw.Stop();
            stpw.Reset();
            stpw.Start();
        }

        public static string StopTimer()
        {
            stpw.Stop();
            intTimerHits += 1;
            string Results = intTimerHits + "  Stopwatch: MS:" + stpw.ElapsedMilliseconds + " Ticks: " + stpw.ElapsedTicks;
            Debug.Print(Results);
            return Results;
        }

        public static string ElapTime()
        {
            string results = intTimerHits + "  Elapsed: MS:" + stpw.ElapsedMilliseconds + " Ticks: " + stpw.ElapsedTicks;
            Debug.Print(results);
            return results;
        }

        public static void EndProgram()
        {
            GlobalSwitches.ProgramEnding = true;
            Logging.Logger("Ending Program...");
            PurgeTempDir();
        }

        public static void PurgeTempDir()
        {
            try
            {
                Directory.Delete(Paths.DownloadPath, true);
            }
            catch
            {
            }
        }

        public static void AdjustComboBoxWidth(object sender, EventArgs e)
        {
            var senderComboBox = (ComboBox)sender;
            int width = senderComboBox.DropDownWidth;
            Graphics g = senderComboBox.CreateGraphics();
            Font font = senderComboBox.Font;
            int vertScrollBarWidth = (senderComboBox.Items.Count > senderComboBox.MaxDropDownItems) ? SystemInformation.VerticalScrollBarWidth : 0;
            int newWidth = 0;
            foreach (string s in ((ComboBox)sender).Items)
            {
                newWidth = Convert.ToInt32(g.MeasureString(s, font).Width) + vertScrollBarWidth;
                if (width < newWidth)
                {
                    width = newWidth;
                }
            }
            senderComboBox.DropDownWidth = width;
        }

        public static string NotePreview(string Note, int CharLimit = 50)
        {
            if (!string.IsNullOrEmpty(Note))
            {
                if (Note.Length > CharLimit)
                {
                    return Note.Substring(0, CharLimit) + "...";
                }
                else
                {
                    return Note;
                }
            }
            else
            {
                return "";
            }
        }

        public static DialogResult Message(string Prompt, int Buttons = (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Information, string Title = null, Form ParentFrm = null)
        {
            SetWaitCursor(false, ParentFrm);
            AdvancedDialog NewMessage = new AdvancedDialog(ParentFrm);
            return NewMessage.DialogMessage(Prompt, Buttons, Title, ParentFrm);
        }
        
        public static bool OKToEnd()
        {
            if (GlobalSwitches.BuildingCache)
            {
                Message("Still building DB Cache. Please wait and try again.", (int)MessageBoxButtons.OK + (int)MessageBoxIcon.Information, "Critical Function Running");
                return false;
            }

            var GKUpdInstance = Helpers.ChildFormControl.GKUpdaterInstance();
            if (GKUpdInstance.Visible && !GKUpdInstance.OKToClose())
                return false;
            return true;

        }

        public delegate void SetWaitCursorVoidDelegate(bool waiting, Form parentForm = null);
        public static void SetWaitCursor(bool Waiting, Form parentForm = null)
        {
            if (parentForm == null)
            {
                Application.UseWaitCursor = Waiting;
                Application.DoEvents();
            }
            else
            {
                if (parentForm.InvokeRequired)
                {
                    SetWaitCursorVoidDelegate d = new SetWaitCursorVoidDelegate(SetWaitCursor);
                    parentForm.Invoke(d, new object[] { Waiting, parentForm });
                }
                else
                {
                    if (Waiting)
                    {
                        parentForm.Cursor = Cursors.AppStarting;
                    }
                    else
                    {
                        parentForm.Cursor = Cursors.Default;
                    }
                    parentForm.Update();
                }
            }
        }

        public static string RTFToPlainText(string rtfText)
        {
            try
            {
                if (rtfText.StartsWith("{\\rtf"))
                {
                    using (RichTextBox rtBox = new RichTextBox())
                    {
                        rtBox.Rtf = rtfText;
                        return rtBox.Text;
                    }
                }
                else
                {
                    return rtfText;
                }
            }
            catch (ArgumentException ex)
            {
                //If we get an argument error, that means the text is not RTF so we return the plain text.
                return rtfText;
            }
        }

        public static void SetRichTextBox(RichTextBox richTextBox, string text)
        {
            if (text.StartsWith("{\\rtf"))
            {
                richTextBox.Rtf = text;
            }
            else
            {
                richTextBox.Text = text;
            }
        }

    }
}
